File: Android/shared/src/main/kotlin/shared/dto/objects/finance/Cashbox.kt
@file:UseSerializers(
    UUIDSerializer::class,
    DateISOSerializer::class
)


package shared.dto.objects.finance

import shared.serializers.DateISOSerializer
import kotlinx.serialization.Serializable
import kotlinx.serialization.UseSerializers
import shared.dto.enums.PaymentType
import shared.dto.protocols.Parametable
import shared.dto.protocols.Responsable
import shared.serializers.UUIDSerializer
import java.util.*

object Cashbox {
    data object Parameters {
        @Serializable
        data class Create(
            val salonId: UUID,
            val paymentType: PaymentType,
        ) : Parametable()

        @Serializable
        data class Retrieve(
            val paymentType: PaymentType?,
            val startDate: Date?,
            val endDate: Date?,
        ) : Parametable()
    }

    data object Responses {
        data class Full(
            val id: UUID,
            val createDate: Date,
            val paymentType: PaymentType,
        ) : Responsable
    }
}

File: Node/src/dto/objects/finance/Cashbox.ts
import { PaymentType } from "../../enums/PaymentType"

// TODO: 
// FLOW EXAMPLE
//import * as Cashbox from "./Cashbox";
//
// const cashbox: Cashbox.Create = {
//     salonId: "",
//     paymentType: PaymentType.BANK_TRANSFER,
// }

// const createDTO: Create = cashbox

export type Cashbox = Parameters | Responses

type Parameters = Create | Retrieve

type Responses = Full

export interface Create {
    salonId: string // TODO: UUID представлено как строка??
    paymentType: PaymentType
}

export interface Retrieve {
    paymentType: PaymentType | undefined
    startDate: Date | undefined
    endDate: Date | undefined
}

export interface Full {
    id: string // TODO: UUID представлено как строка??
    createDate: Date
    paymentType: PaymentType
}